engines/cryomni3d/versailles/engine.cpp

```diff
@@ -0,0 +1,1443 @@
+
+	unsigned int srcState = _placeStates[srcPlaceId].state;
+	unsigned int dstState = _placeStates[dstPlaceId].state;
+
+	if (srcState >= srcNumStates) {
+		error("Invalid src state");
+	}
+
+	if (dstState >= dstNumStates) {
+		error("Invalid dst state");
+	}
+
+	if (animsNum <= 1) {
+		return 0;
+	}
+
+	if (srcNumStates == 2 && dstNumStates == 2) {
+		if (animsNum == 2) {
+			return dstState;
+		} else if (animsNum == 4) {
+			return srcState * 2 + dstState;
+		}
+	}
+
+	if (animsNum == dstNumStates) {
+		return dstState;
+	}
+
+	if (animsNum == srcNumStates) {
+		return srcState;
+	}
+
+	// Other case
+	return 0;
+}
+
+int CryOmni3DEngine_Versailles::handleWarp() {
+	bool exit = false;
+	bool leftButtonPressed = false;
+	bool firstDraw = true;
+	bool moving = true;
+	unsigned int actionId;
+	g_system->showMouse(true);
+	while (!leftButtonPressed && !exit) {
+		int xDelta = 0, yDelta = 0;
+		unsigned int movingCursor = -1;
+
+		pollEvents();
+		Common::Point mouse = getMousePos();
+
+		if (mouse.y < 100) {
+			movingCursor = 245;
+			yDelta = 100 - mouse.y;
+		} else if (mouse.y > 380) {
+			movingCursor = 224;
+			yDelta = 380 - mouse.y;
+		}
+		if (mouse.x < 100) {
+			movingCursor = 241;
+			xDelta = 100 - mouse.x;
+		} else if (mouse.x > 540) {
+			movingCursor = 228;
+			xDelta = 540 - mouse.x;
+		}
+		if (_omni3dSpeed > 0) {
+			xDelta <<= _omni3dSpeed;
+			yDelta <<= _omni3dSpeed;
+		} else if (_omni3dSpeed < 0) {
+			xDelta >>= -_omni3dSpeed;
+			yDelta >>= -_omni3dSpeed;
+		}
+		leftButtonPressed = (getCurrentMouseButton() == 1);
+
+		Common::Point mouseRev = _omni3dMan.mapMouseCoords(mouse);
+		mouseRev.y = 768 - mouseRev.y;
+
+		actionId = _currentPlace->hitTest(mouseRev);
+
+		exit = handleWarpMouse(&actionId, movingCursor);
+		if (g_engine->shouldQuit()) {
+			_abortCommand = AbortQuit;
+			exit = true;
+		}
+		if (exit) {
+			actionId = 66666;
+		}
+
+		if (firstDraw || xDelta || yDelta || _omni3dMan.hasSpeed()) {
+			bool useOldSpeed = false;
+			if (_omni3dSpeed <= 2) {
+				useOldSpeed = true;
+			}
+			_omni3dMan.updateCoords(xDelta, -yDelta, useOldSpeed);
+
+			const Graphics::Surface *result = _omni3dMan.getSurface();
+			g_system->copyRectToScreen(result->getPixels(), result->pitch, 0, 0, result->w, result->h);
+			if (!exit) {
+				// TODO: countdown
+				g_system->updateScreen();
+				if (firstDraw && _fadedPalette) {
+					fadeInPalette(_mainPalette);
+					_fadedPalette = false;
+				}
+			}
+			moving = true;
+			firstDraw = false;
+		} else if (moving) {
+			const Graphics::Surface *result = _omni3dMan.getSurface();
+			g_system->copyRectToScreen(result->getPixels(), result->pitch, 0, 0, result->w, result->h);
+			if (!exit) {
+				// TODO: countdown
+				g_system->updateScreen();
+			}
+			// TODO: cursorUseZones
+			moving = false;
+		} else {
+			if (!exit) {
+				// TODO: countdown
+				g_system->updateScreen();
+			}
+		}
+		if (!exit && !leftButtonPressed) {
+			g_system->delayMillis(50);
```

Comment:
Is there any reason for using such a long delay? It keeps the framerate low; the game could run very smoothly on modern systems.

Reply:
I put that at the beginning of the development.
The original game doesn't have this sleep time and it makes all movements way too fast.
In the same way, the original game uses CPU at 100% which this delay prevents.

Maybe it can be decreased but that will mean to adjust Omni3D movements at least.

What do you mean when you say that the framerate is low? Do you mean that the fps figure is low or that movements are slow?

Reply:
I mean the frames per second figure is low. On my system, the game runs at 15 fps. Mouse movement is not very smooth, and neither is the Omni3D movement. Perhaps the delay could be reduced and the Omni3D movement speed computed relative to the game speed to have framerate independence.

Reply:
I see what you mean and I agree. This just needs some thinking and checks.
I am not sure if there are other parts of the code which depends on this wait.
I will try to see that next week-end.